internal/deb/extract.go

```diff
@@ -277,6 +282,140 @@ func extractData(dataReader io.Reader, options *ExtractOptions) error {
 	return nil
 }
 
+type pendingHardLink struct {
+	path         string
+	extractInfos []ExtractInfo
+}
+
+type extractHardLinkOptions struct {
+	*ExtractOptions
+	pendingLinks map[string][]pendingHardLink
+}
+
+// extractHardLinks iterates through the tarball a second time to extract the
+// hard links that were not extracted in the first pass.
+func extractHardLinks(pkgReader io.ReadSeeker, opts *extractHardLinkOptions) error {
+	offset, err := pkgReader.Seek(0, io.SeekStart)
+	if err != nil {
+		return err
+	}
+	if offset != 0 {
+		return fmt.Errorf("internal error: cannot seek to the beginning of the package")
+	}
+	dataReader, err := getDataReader(pkgReader)
+	if err != nil {
+		return err
+	}
+	defer dataReader.Close()
+
+	tarReader := tar.NewReader(dataReader)
+	for {
+		tarHeader, err := tarReader.Next()
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			return err
+		}
+
+		sourcePath := sanitizeTarPath(tarHeader.Name)
+		if sourcePath == "" {
+			continue
+		}
+
+		links, ok := opts.pendingLinks[sourcePath]
+		if !ok || len(links) == 0 {
```

Comment:
FWIW, in this particular case the `ok` may be safely omitted:
```
links := opts.pendingLinks[sourcePath]
if len(links) == 0 {
        continue
}
```

Reply:
Done in https://github.com/canonical/chisel/pull/179/commits/801f3c5dcf187edd019211ed422a466176f09562.